# 📊 ANOVA Test in Statistics
### Analysis of Variance — Comparing More Than Two Groups
---

> **ANOVA (Analysis of Variance)** is used to determine whether there are **statistically significant differences between the means of 3 or more independent groups**.  
> When we have more than 2 groups, running multiple T-Tests increases error — ANOVA solves this in **one single test**.

---

## 🤔 T-Test vs ANOVA — What's the Difference?

| | T-Test | ANOVA |
|---|---|---|
| Number of Groups | 2 | 3 or more |
| What it compares | 2 means | Multiple means simultaneously |
| Risk of Error | Low | T-Test on 3+ groups inflates error — ANOVA fixes this |
| Test Statistic | T-Score | F-Statistic |

---

## 📌 Problem Statement

We use the **Titanic Dataset** to answer:

> **"Is there a significant difference in the Ages of passengers across the 3 Passenger Classes (1st, 2nd, 3rd)?"**

- **Variable 1 (Numerical):** Age  
- **Variable 2 (Categorical):** Passenger Class (1, 2, 3)  

---

## 🔢 Formula

$$F = \frac{\text{Variance Between Groups}}{\text{Variance Within Groups}}$$

- **High F-Statistic** → Means are far apart → Groups are significantly different  
- **Low F-Statistic** → Means are close → No significant difference  

---

## 🧪 Our Hypotheses

- **H₀ (Null Hypothesis):** Mean age is **equal** across all 3 passenger classes
- **H₁ (Alternative Hypothesis):** At least **one class** has a significantly different mean age

In [1]:
# Importing required libraries
import seaborn as sns
import pandas as pd
from scipy.stats import f_oneway

## 📦 Step 1 — Load the Titanic Dataset

We load the Titanic dataset and keep only the columns we need — `age` and `pclass`. We also drop any missing age values.

In [2]:
df = sns.load_dataset('titanic')

# Keep only relevant columns and drop missing age values
df = df[['age', 'pclass']].dropna()

print(f"Dataset Shape after cleaning : {df.shape}")
print(f"Passenger Classes            : {df['pclass'].unique()}")
print(f"Total Passengers with Age    : {len(df)}")
print("-" * 40)
df.head(10)

Dataset Shape after cleaning : (714, 2)
Passenger Classes            : [3 1 2]
Total Passengers with Age    : 714
----------------------------------------


,age,pclass
0,22.0,3
1,38.0,1
2,26.0,3
3,35.0,1
4,35.0,3
6,54.0,1
7,2.0,3
8,27.0,3
9,14.0,2
10,4.0,3


## 📋 Step 2 — Split Data into 3 Groups by Passenger Class

We separate the age data for each of the 3 passenger classes so we can compare them individually.

In [4]:
class_1 = df[df['pclass'] == 1]['age']
class_2 = df[df['pclass'] == 2]['age']
class_3 = df[df['pclass'] == 3]['age']

print("-" * 45)
print(f"{'Class':<12} {'Count':<10} {'Mean Age':<12} {'Std Dev'}")
print("-" * 45)
print(f"{'1st Class':<12} {len(class_1):<10} {class_1.mean():<12.2f} {class_1.std():.2f}")
print(f"{'2nd Class':<12} {len(class_2):<10} {class_2.mean():<12.2f} {class_2.std():.2f}")
print(f"{'3rd Class':<12} {len(class_3):<10} {class_3.mean():<12.2f} {class_3.std():.2f}")
print("-" * 45)

---------------------------------------------
Class        Count      Mean Age     Std Dev
---------------------------------------------
1st Class    186        38.23        14.80
2nd Class    173        29.88        14.00
3rd Class    355        25.14        12.50
---------------------------------------------


## 🔍 Step 3 — Perform the ANOVA Test

We use `f_oneway()` from SciPy which performs a **One-Way ANOVA** test.

`f_oneway()` takes the data of each group separately and returns:

| Output | Meaning |
|---|---|
| `f_stats` | F-Statistic — ratio of between-group variance to within-group variance |
| `p_value` | Probability that the differences happened by random chance |

> A **very high F-Statistic** means the group means are very far apart relative to the variation within each group — strong evidence against H₀.

In [5]:
f_stats, p_value = f_oneway(class_1, class_2, class_3)

print(f"F-Statistic : {f_stats:.4f}")
print(f"P-Value     : {p_value:.2e}")

F-Statistic : 57.4435
P-Value     : 7.49e-24


## ✅ Step 4 — Decision: Accept or Reject Null Hypothesis?

### Decision Rule:

| Condition | Decision |
|---|---|
| p_value < α (0.05) | ❌ Reject H₀ — at least one group mean is different |
| p_value ≥ α (0.05) | ✅ Accept H₀ — no significant difference between groups |

In [7]:
alpha = 0.05

print(f"F-Statistic : {f_stats:.4f}")
print(f"P-Value     : {p_value:.2e}")
print(f"Alpha       : {alpha}")
print("-" * 50)

if p_value < alpha:
    print(" Reject the Null Hypothesis")
    print("→ There IS a significant difference in age across at least one passenger class.")
else:
    print(" Accept the Null Hypothesis")
    print("→ No significant difference in age across passenger classes.")

F-Statistic : 57.4435
P-Value     : 7.49e-24
Alpha       : 0.05
--------------------------------------------------
 Reject the Null Hypothesis
→ There IS a significant difference in age across at least one passenger class.


## 📝 Conclusion

| Metric | Value |
|---|---|
| Dataset | Titanic (714 passengers with age) |
| Groups | 1st Class, 2nd Class, 3rd Class |
| 1st Class Mean Age | ~38.2 years |
| 2nd Class Mean Age | ~29.9 years |
| 3rd Class Mean Age | ~25.1 years |
| F-Statistic | 57.4435 |
| P-Value | 7.49e-24 (~0) |
| Alpha (α) | 0.05 |
| Decision | ❌ Reject H₀ |

---

**Interpretation:**  
With an **F-Statistic of 57.44** and a **p-value of 7.49e-24 (essentially zero)**, we strongly reject the null hypothesis.  
The age distribution across the 3 passenger classes is **significantly different** — 1st class passengers were notably older than 2nd and 3rd class passengers.

> This makes real-world sense — wealthier, older passengers could afford 1st class tickets, while younger passengers traveled in 3rd class.

---

## 🔁 All Five Tests — Final Comparison

| Test | Data Type | Groups | Our Question | Decision |
|---|---|---|---|---|
| Z-Test | Numerical | 1 vs Population | Height = 170cm? | ✅ Accept H₀ |
| One Sample T-Test | Numerical | 1 vs Population | Height = 170cm? | ✅ Accept H₀ |
| Two Sample T-Test | Numerical | 2 Groups | Group A = Group B? | ❌ Reject H₀ |
| Chi-Square Test | Categorical | 2 Categories | Gender affects Survival? | ❌ Reject H₀ |
| **ANOVA Test** | **Numerical** | **3+ Groups** | **Age differs by Class?** | **❌ Reject H₀** |

> 💡 **Key Takeaway:** Use **ANOVA** whenever you want to compare means across **3 or more groups** simultaneously. Never run multiple T-Tests for this — ANOVA is the right tool. A high F-Statistic with a near-zero p-value means the group differences are real and statistically significant.

